### PR TITLE
III-6220 Refactor ManagementToken as preparation for Keycloak

### PR DESCRIPTION
**Pull request by LucWollants**
### Changed
- Refactor `Auth0` specific management code to interfaces so they can be implemented with a Keycloak flavor
- The `JsonDocumentFetcher` only needs a token
- The `AuthenticateRequest` also needs to make a management call

### Note
- The behavior should be the same as before. But with these changes and interfaces, it is possible to implement a Keycloak flavor
 
---

Ticket: https://jira.uitdatabank.be/browse/III-6220

### CODE REVIEW

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+use CultuurNet\UDB3\Search\Json;
+use DateTimeImmutable;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Psr7\Request;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Psr\Log\NullLogger;
+
+final class Auth0ManagementTokenGenerator implements ManagementTokenGenerator, LoggerAwareInterface
```
Comment (grubolsch) [inline src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:18]: Is there any way to set the Logger? It should be an optional dependency?

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    ) {
+        $this->domain = $domain;
+        $this->client = $client;
+        $this->clientId = $clientId;
+        $this->clientSecret = $clientSecret;
+        $this->audience = $audience;
+        $this->logger = new NullLogger();
+    }
+
+    public function newToken(): ManagementToken
```
Comment (grubolsch) [inline src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:47]: I personally like the name functions that make api calls with the prefix "fetch". This fetches a different token each time you call it, it does not return the same "new token".
So I would rename it to fetchNewToken()

File: src/Http/Authentication/ManagementToken/ManagementTokenProvider.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ) {
+        $this->tokenGenerator = $tokenGenerator;
+        $this->tokenRepository = $tokenRepository;
+    }
+
+    public function token(): string
+    {
+        $token = $this->tokenRepository->get();
+
+        if ($token === null || $this->expiresWithin($token, '+5 minutes')) {
```
Comment (grubolsch) [inline src/Http/Authentication/ManagementToken/ManagementTokenProvider.php:27]: Move string to a constant

File: src/Http/Authentication/ManagementToken/ManagementTokenProvider.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            $token = $this->tokenGenerator->newToken();
+            $this->tokenRepository->set($token);
+        }
+
+        return $token->getToken();
+    }
+
+    private function expiresWithin(ManagementToken $token, string $offset): bool
```
Comment (grubolsch) [inline src/Http/Authentication/ManagementToken/ManagementTokenProvider.php:35]: offset will always be +5 minutes so remove param and use constant/string hardcoded?

File: src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php [hunk 5]
```diff
@@ -16,22 +16,22 @@ public function __construct(string $fullFilePath)
...
+    public function get(): ?ManagementToken
     {
         if (!file_exists($this->fullFilePath)) {
             return null;
         }
 
         $tokenAsArray = Json::decodeAssociatively(file_get_contents($this->fullFilePath));
 
-        return new Auth0Token(
+        return new ManagementToken(
```
Comment (grubolsch) [inline src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php:27]: Should we check if the fields `token`, `issuesAt` & `expiresIn` are returned? If not return `NULL`?

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                throw new ConnectException(
+                    $message,
+                    new Request('POST', 'https://' . $this->domain . '/oauth/token')
+                );
+            }
+
+            $this->logger->info($message);
+        }
+
+        $res = Json::decodeAssociatively($response->getBody()->getContents());
```
Comment (grubolsch) [inline src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:76]: I think this is a difference in style, but I would check if the fields `access_token` & `expires_in` are returned.
And check if JSON throws a JsonException, in either of these cases I would throw a ConnectException.

**Review by grubolsch — CHANGES_REQUESTED** [review]

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php (hunk 1 above)
Comment (LucWollants) [reply src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:18]: It works with the `LoggerAwareTrait` trait, but it was never set before. We could create a follow-up ticket to change this behavior.

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php (hunk 2 above)
Comment (LucWollants) [reply src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:47]: The method comes from the interface `ManagementTokenGenerator`, this interface does not imply that the token is created through an HTTP request, the token could also be created locally. Maybe `generate` is a better name that matches the `ManagementTokenGenerator` interface better?

File: src/Http/Authentication/ManagementToken/ManagementTokenProvider.php (hunk 4 above)
Comment (LucWollants) [reply src/Http/Authentication/ManagementToken/ManagementTokenProvider.php:35]: I changed the method to `isExpired` and the offset is internal, so no longer need for a constant

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php (hunk 6 above)
Comment (LucWollants) [reply src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:76]: That means there was something wrong with the `set` method and we would never notice this. If we overlook that and always return `null` from the cache we will always create a new token. So I would prefer a hard crash that allows us to fix the issue.

File: src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php (hunk 5 above)
Comment (LucWollants) [reply src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php:27]: That means there was something wrong with the set method and we would never notice this. If we overlook that and always return null from the cache we will always create a new token. So I would prefer a hard crash that allows us to fix the issue.

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php (hunk 2 above)
Comment (grubolsch) [reply src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:47]: I can live with generate :-)

File: src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php (hunk 5 above)
Comment (grubolsch) [reply src/Http/Authentication/ManagementToken/ManagementTokenFileRepository.php:27]: okay

**Review by grubolsch — APPROVED** [review]

File: src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php (hunk 2 above)
Comment (LucWollants) [reply src/Http/Authentication/Auth0/Auth0ManagementTokenGenerator.php:47]: I became a little more complex. Turns out we need to create a management token to validate an incoming access token an we need to create a login token to make authenticated requests to UDB3. So in the follow-up pull request the interface changed again 😅

**Pull request merged**